include/vast/CodeGen/CodeGenAttrVisitor.hpp

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2023-present, Trail of Bits, Inc.
+
+#pragma once
+
+#include "vast/Util/Warnings.hpp"
+
+VAST_RELAX_WARNINGS
+#include <clang/AST/AttrVisitor.h>
+#include <clang/AST/Attr.h>
+#include <clang/Basic/Diagnostic.h>
+#include <clang/Frontend/FrontendDiagnostic.h>
+VAST_UNRELAX_WARNINGS
+
+#include <mlir/IR/Attributes.h>
+#include <vast/CodeGen/Types.hpp>
+
+namespace vast::cg {
+
+    template< typename Derived >
+    struct CodeGenAttrVisitor
+        : clang::ConstAttrVisitor< CodeGenAttrVisitor< Derived >, operation >
+        , CodeGenVisitorLens< CodeGenAttrVisitor< Derived >, Derived >
+        , CodeGenBuilder< CodeGenAttrVisitor< Derived >, Derived >
+    {
+        using LensType = CodeGenVisitorLens< CodeGenAttrVisitor< Derived >, Derived >;
+
+        using LensType::derived;
+        using LensType::context;
+        using LensType::mcontext;
+        using LensType::acontext;
+
+        using LensType::visit;
+
+        using LensType::meta_location;
+
+        using Builder = CodeGenBuilder< CodeGenAttrVisitor< Derived >, Derived >;
+
+        using Builder::builder;
+
+        template< typename Attr, typename... Args >
+        auto make(Args &&...args) {
+            return builder().template getAttr< Attr >(std::forward< Args >(args)...);
+        }
+
+        auto Visit(const clang::Attr *attr) -> mlir_attr {
```

Comment:
this should not be needed `ConstAttrVisitor` should provide the default impl that does the same?